site/en/docs/extensions/mv3/promises/index.md

```diff
@@ -0,0 +1,214 @@
+{% endif %}
+
+## How to use promises
+
+There are many places where using promises will result in cleaner, easier-to-maintain code. You
+should consider using promises in situations such as the following:
+
+* Any time that you want to clean up your code by using a more "synchronous" invocation style.
+* Where error handling would be too difficult using callbacks.
+* When you want a simpler way to invoke a number of concurrent methods and gather the results into a single thread of code.
+
+### Converting a callback to a promise {: #compare-to-callback}
+
+One way to understand how you can use promises in extensions APIs is to compare two equivalent code
+fragments, one using a callback and one using a promise. The following example shows this
+comparison:
+
+<!--
+// --- Standard callback implementation ---
+// --- Promise implementation ---
+-->
+
+#### Standard callback implementation
+
+```js
+
+function openTabOnRight(onComplete) {
+  chrome.tabs.query(queryOptions, function(tabs) {
+    if (chrome.runtime.lastError) {
+      onComplete({error: chrome.runtime.lastError});
+      return;
+    }
+
+    if (!tabs.length) {
+      onComplete();
+      return;
+    };
+
+    chrome.tabs.create({
+      url: 'https://example.com',
+      index: tab[0].index + 1,
+    }, function(tab) {
+      if (chrome.runtime.lastError) {
+        onComplete({error: chrome.runtime.lastError});
+      }
+
+      console.log('tab created', tab);
+      onComplete(tab);
+    });
+  });
+}
+```
+
+#### Promise implementation
+
+```js
+function openTabOnRight() {
+  // Errors are automatically propagated down the promise chain
+  return chrome.tabs.query(queryOptions)
+    .then((tabs) => {
+      if (!tabs.length) return;
+
+      return chrome.tabs.create({
+        url: 'https://example.com',
+        index: tab[0].index + 1,
+      });
+    })
+    .then(tab => {
+      if (!tab) return;
+
+      console.log('tab created', tab);
+      return tab;
+    });
+}
+```
+
+
+<!-- notes here about the code fragments -->
+
+### Error handling
+
+Returning errors works differently depending on if the extension is using a callback or a promise.
+
+#### Error handling with Callbacks
+
+If using a callback, then `chrome.runtime.lastError` is set for the duration of the execution of the
+callback. It is not thrown as a JS Error (which would interrupt JS execution), and is not set
+outside the duration of the callback run (which would result in it being "randomly" set during other
+execution).  The extension would look at the last error like this:
+
+```js
+chrome.tabs.create({...}, (result) => {
+  if (chrome.runtime.lastError) {
+    // Handle last error
+  }
+});
+```
+
+Promises are designed to deliver asynchronous results, both success and failure.  A failure in a
+promise (a promise rejection) is handled differently.  It might look like this:
+
+```js
+chrome.tabs.create({...})
+  .then((result) => {
+    // success case
+  })
+  .catch((e) => {
```

Comment:
```suggestion
  .catch((error) => {
```

Rename for clarity. While 'e' or 'err' are often used for error variable names, I think it's best to err on the side of being explicit here.